// posthog-js/lib/src/decide.js
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
import { autocapture } from './autocapture';
import { loadScript } from './utils';
import { Compression } from './types';
import { STORED_GROUP_PROPERTIES_KEY, STORED_PERSON_PROPERTIES_KEY } from './constants';
import { _isUndefined } from './utils/type-utils';
import { logger } from './utils/logger';
import { window, document, assignableWindow } from './utils/globals';
var Decide = /** @class */ (function () {
    function Decide(instance) {
        this.instance = instance;
        // don't need to wait for `decide` to return if flags were provided on initialisation
        this.instance.decideEndpointWasHit = this.instance._hasBootstrappedFeatureFlags();
    }
    Decide.prototype.call = function () {
        var _this = this;
        /*
        Calls /decide endpoint to fetch options for autocapture, session recording, feature flags & compression.
        */
        var data = {
            token: this.instance.config.token,
            distinct_id: this.instance.get_distinct_id(),
            groups: this.instance.getGroups(),
            person_properties: this.instance.get_property(STORED_PERSON_PROPERTIES_KEY),
            group_properties: this.instance.get_property(STORED_GROUP_PROPERTIES_KEY),
            disable_flags: this.instance.config.advanced_disable_feature_flags ||
                this.instance.config.advanced_disable_feature_flags_on_first_load ||
                undefined,
        };
        this.instance._send_request({
            method: 'POST',
            url: this.instance.requestRouter.endpointFor('api', '/decide/?v=3'),
            data: data,
            compression: this.instance.config.disable_compression ? undefined : Compression.Base64,
            timeout: this.instance.config.feature_flag_request_timeout_ms,
            callback: function (response) { return _this.parseDecideResponse(response.json); },
        });
    };
    Decide.prototype.parseDecideResponse = function (response) {
        var e_1, _a;
        var _this = this;
        var _b;
        this.instance.featureFlags.setReloadingPaused(false);
        // :TRICKY: Reload - start another request if queued!
        this.instance.featureFlags._startReloadTimer();
        var errorsLoading = !response;
        if (!this.instance.config.advanced_disable_feature_flags_on_first_load &&
            !this.instance.config.advanced_disable_feature_flags) {
            this.instance.featureFlags.receivedFeatureFlags(response !== null && response !== void 0 ? response : {}, errorsLoading);
        }
        if (errorsLoading) {
            logger.error('Failed to fetch feature flags from PostHog.');
            return;
        }
        if (!(document && document.body)) {
            logger.info('document not ready yet, trying again in 500 milliseconds...');
            setTimeout(function () {
                _this.parseDecideResponse(response);
            }, 500);
            return;
        }
        this.instance.toolbar.afterDecideResponse(response);
        (_b = this.instance.sessionRecording) === null || _b === void 0 ? void 0 : _b.afterDecideResponse(response);
        autocapture.afterDecideResponse(response, this.instance);
        this.instance._afterDecideResponse(response);
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        var surveysGenerator = window === null || window === void 0 ? void 0 : window.extendPostHogWithSurveys;
        if (response['surveys'] && !surveysGenerator) {
            loadScript(this.instance.requestRouter.endpointFor('assets', '/static/surveys.js'), function (err) {
                if (err) {
                    return logger.error("Could not load surveys script", err);
                }
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                window.extendPostHogWithSurveys(_this.instance);
            });
        }
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        var exceptionAutoCaptureAddedToWindow = window === null || window === void 0 ? void 0 : window.extendPostHogWithExceptionAutoCapture;
        if (response['autocaptureExceptions'] &&
            !!response['autocaptureExceptions'] &&
            _isUndefined(exceptionAutoCaptureAddedToWindow)) {
            loadScript(this.instance.requestRouter.endpointFor('assets', '/static/exception-autocapture.js'), function (err) {
                if (err) {
                    return logger.error("Could not load exception autocapture script", err);
                }
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                window.extendPostHogWithExceptionAutocapture(_this.instance, response);
            });
        }
        if (response['siteApps']) {
            if (this.instance.config.opt_in_site_apps) {
                var _loop_1 = function (id, url) {
                    var scriptUrl = this_1.instance.requestRouter.endpointFor('api', url);
                    assignableWindow["__$$ph_site_app_".concat(id)] = this_1.instance;
                    loadScript(scriptUrl, function (err) {
                        if (err) {
                            logger.error("Error while initializing PostHog app with config id ".concat(id), err);
                        }
                    });
                };
                var this_1 = this;
                try {
                    for (var _c = __values(response['siteApps']), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var _e = _d.value, id = _e.id, url = _e.url;
                        _loop_1(id, url);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            else if (response['siteApps'].length > 0) {
                logger.error('PostHog site apps are disabled. Enable the "opt_in_site_apps" config to proceed.');
            }
        }
    };
    return Decide;
}());
export { Decide };
