// posthog-js/lib/src/request.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import { _base64Encode, _each } from './utils';
import Config from './config';
import { Compression } from './types';
import { _formDataToQuery } from './utils/request-utils';
import { logger } from './utils/logger';
import { fetch, document, XMLHttpRequest, AbortController, navigator } from './utils/globals';
import { gzipSync, strToU8 } from 'fflate';
// eslint-disable-next-line compat/compat
export var SUPPORTS_REQUEST = !!XMLHttpRequest || !!fetch;
var CONTENT_TYPE_PLAIN = 'text/plain';
var CONTENT_TYPE_JSON = 'application/json';
var CONTENT_TYPE_FORM = 'application/x-www-form-urlencoded';
// This is the entrypoint. It takes care of sanitizing the options and then calls the appropriate request method.
export var request = function (_options) {
    // Clone the options so we don't modify the original object
    var options = __assign({}, _options);
    options.timeout = options.timeout || 60000;
    options.url = extendURLParams(options.url, {
        _: new Date().getTime().toString(),
        ver: Config.LIB_VERSION,
        compression: options.compression,
    });
    if (options.transport === 'sendBeacon' && (navigator === null || navigator === void 0 ? void 0 : navigator.sendBeacon)) {
        return _sendBeacon(options);
    }
    // NOTE: Until we are confident with it, we only use fetch if explicitly told so
    // At some point we will make it the default over xhr
    if (options.transport === 'fetch' && fetch) {
        return _fetch(options);
    }
    if (XMLHttpRequest || !document) {
        return xhr(options);
    }
    // Final fallback if everything else fails...
    scriptRequest(options);
};
export var extendURLParams = function (url, params) {
    var _a = __read(url.split('?'), 2), baseUrl = _a[0], search = _a[1];
    var newParams = __assign({}, params);
    search === null || search === void 0 ? void 0 : search.split('&').forEach(function (pair) {
        var _a = __read(pair.split('='), 1), key = _a[0];
        delete newParams[key];
    });
    var newSearch = _formDataToQuery(newParams);
    newSearch = newSearch ? (search ? search + '&' : '') + newSearch : search;
    return "".concat(baseUrl, "?").concat(newSearch);
};
var encodeToDataString = function (data) {
    return 'data=' + encodeURIComponent(typeof data === 'string' ? data : JSON.stringify(data));
};
var encodePostData = function (_a) {
    var data = _a.data, compression = _a.compression;
    if (!data) {
        return;
    }
    if (compression === Compression.GZipJS) {
        var gzipData = gzipSync(strToU8(JSON.stringify(data)), { mtime: 0 });
        return {
            contentType: CONTENT_TYPE_PLAIN,
            body: new Blob([gzipData], { type: CONTENT_TYPE_PLAIN }),
        };
    }
    if (compression === Compression.Base64) {
        var b64data = _base64Encode(JSON.stringify(data));
        return {
            contentType: CONTENT_TYPE_FORM,
            body: encodeToDataString(b64data),
        };
    }
    return {
        contentType: CONTENT_TYPE_JSON,
        body: JSON.stringify(data),
    };
};
var xhr = function (options) {
    var _a;
    var req = new XMLHttpRequest();
    req.open(options.method || 'GET', options.url, true);
    var _b = (_a = encodePostData(options)) !== null && _a !== void 0 ? _a : {}, contentType = _b.contentType, body = _b.body;
    _each(options.headers, function (headerValue, headerName) {
        req.setRequestHeader(headerName, headerValue);
    });
    if (contentType) {
        req.setRequestHeader('Content-Type', contentType);
    }
    if (options.timeout) {
        req.timeout = options.timeout;
    }
    // send the ph_optout cookie
    // withCredentials cannot be modified until after calling .open on Android and Mobile Safari
    req.withCredentials = true;
    req.onreadystatechange = function () {
        var _a;
        // XMLHttpRequest.DONE == 4, except in safari 4
        if (req.readyState === 4) {
            var response = {
                statusCode: req.status,
                text: req.responseText,
            };
            if (req.status === 200) {
                try {
                    response.json = JSON.parse(req.responseText);
                }
                catch (e) {
                    // logger.error(e)
                }
            }
            (_a = options.callback) === null || _a === void 0 ? void 0 : _a.call(options, response);
        }
    };
    req.send(body);
};
var _fetch = function (options) {
    var _a;
    var _b = (_a = encodePostData(options)) !== null && _a !== void 0 ? _a : {}, contentType = _b.contentType, body = _b.body;
    // eslint-disable-next-line compat/compat
    var headers = new Headers();
    _each(headers, function (headerValue, headerName) {
        headers.append(headerName, headerValue);
    });
    if (contentType) {
        headers.append('Content-Type', contentType);
    }
    var url = options.url;
    var aborter = null;
    if (AbortController) {
        var controller_1 = new AbortController();
        aborter = {
            signal: controller_1.signal,
            timeout: setTimeout(function () { return controller_1.abort(); }, options.timeout),
        };
    }
    fetch(url, {
        method: (options === null || options === void 0 ? void 0 : options.method) || 'GET',
        headers: headers,
        keepalive: options.method === 'POST',
        body: body,
        signal: aborter === null || aborter === void 0 ? void 0 : aborter.signal,
    })
        .then(function (response) {
        return response.text().then(function (responseText) {
            var _a;
            var res = {
                statusCode: response.status,
                text: responseText,
            };
            if (response.status === 200) {
                try {
                    res.json = JSON.parse(responseText);
                }
                catch (e) {
                    logger.error(e);
                }
            }
            (_a = options.callback) === null || _a === void 0 ? void 0 : _a.call(options, res);
        });
    })
        .catch(function (error) {
        var _a;
        logger.error(error);
        (_a = options.callback) === null || _a === void 0 ? void 0 : _a.call(options, { statusCode: 0, text: error });
    })
        .finally(function () { return (aborter ? clearTimeout(aborter.timeout) : null); });
    return;
};
var _sendBeacon = function (options) {
    // beacon documentation https://w3c.github.io/beacon/
    // beacons format the message and use the type property
    var _a;
    var url = extendURLParams(options.url, {
        beacon: '1',
    });
    try {
        var _b = (_a = encodePostData(options)) !== null && _a !== void 0 ? _a : {}, contentType = _b.contentType, body = _b.body;
        // sendBeacon requires a blob so we convert it
        var sendBeaconBody = typeof body === 'string' ? new Blob([body], { type: contentType }) : body;
        navigator.sendBeacon(url, sendBeaconBody);
    }
    catch (e) {
        // send beacon is a best-effort, fire-and-forget mechanism on page unload,
        // we don't want to throw errors here
    }
};
var scriptRequest = function (options) {
    var _a;
    if (!document) {
        return;
    }
    var script = document.createElement('script');
    script.type = 'text/javascript';
    script.async = true;
    script.defer = true;
    script.src = options.url;
    var s = document.getElementsByTagName('script')[0];
    (_a = s.parentNode) === null || _a === void 0 ? void 0 : _a.insertBefore(script, s);
};
